File: OneDrive.fs
module comicdownloader.OneDrive


open System.Threading.Tasks
open Microsoft.Graph
open System.Net.Http
open System.Collections.Generic
open System.Text
open System.Text.Json.Nodes
open Azure.Core
open System
open System.IO
open Microsoft.Graph.Models
open Microsoft.Extensions.Logging

let client = new HttpClient();

type OneDriveSettings = {
    ClientId: string
    RefreshToken: string
}

let getAccessToken (logger: ILogger) settings =
    task {
        logger.LogInformation("Refreshing OneDrive access token")

        let formValues =
            [
                KeyValuePair<string, string>("client_id", settings.ClientId)
                KeyValuePair<string, string>("refresh_token", settings.RefreshToken)
                KeyValuePair<string, string>("redirect_uri", "http://localhost:8000")
                KeyValuePair<string, string>("grant_type", "refresh_token")
            ]

        use body = new FormUrlEncodedContent(formValues)
        let! result = client.PostAsync("https://login.microsoftonline.com/common/oauth2/v2.0/token", body)
        let! json = result.Content.ReadAsStringAsync()

        if not result.IsSuccessStatusCode then
            logger.LogError("OneDrive token refresh failed with status {StatusCode}", int result.StatusCode)
            return raise (InvalidOperationException($"OneDrive token refresh failed ({int result.StatusCode} {result.StatusCode}): {json}"))
        else
            let node = JsonNode.Parse(json);
            let refreshToken = node["refresh_token"].GetValue<string>()
            let accessToken = node["access_token"].GetValue<string>();

            logger.LogInformation("OneDrive access token refreshed successfully")

            return accessToken, refreshToken
    }

let uploadFile (logger: ILogger) (settings) folderName fileName (bytes:byte array) =
    match settings with
    | None ->
        logger.LogWarning("Skipping OneDrive upload for {FolderName}/{FileName} because no OneDrive settings were found", folderName, fileName)
        Task.FromResult(Error (exn("No settings")))
    | Some settings ->
        task {
            logger.LogInformation("Uploading {FileName} to OneDrive folder {FolderName}", fileName, folderName)

            let! accessToken, _ = getAccessToken logger settings

            let token = DelegatedTokenCredential.Create(fun _ _ -> AccessToken(accessToken, DateTimeOffset.Now.AddMinutes(1)))
            let client = new GraphServiceClient(token)

            try
                use ms = new MemoryStream(bytes)
                let totalLength = bytes.Length
                let! drive = client.Me.Drive.GetAsync()
                let driveId = drive.Id

                let! appRoot = client.Drives[driveId].Special["AppRoot"].GetAsync()
                
                if totalLength > 4 * 1024 * 1024 then
                    let request = client.Drives[driveId].Items[appRoot.Id].ItemWithPath($"{folderName}/{fileName}").CreateUploadSession

                    let uploadSessionRequestBody = 
                        Drives.Item.Items.Item.CreateUploadSession.CreateUploadSessionPostRequestBody(
                            Item = DriveItemUploadableProperties(AdditionalData = dict [ "@microsoft.graph.conflictBehavior", box "replace"])
                        )

                    let! uploadSession = request.PostAsync(uploadSessionRequestBody)
                    
                    let maxSliceSize = 320 * 1024
                    let fileUploadTask = LargeFileUploadTask<DriveItem>(uploadSession, ms, maxSliceSize)

                    let! _ = fileUploadTask.UploadAsync()
                    logger.LogInformation("OneDrive upload completed for {FolderName}/{FileName}", folderName, fileName)
                    return (Ok "File uploaded")
                else
                    let! _ = client.Drives[driveId].Items[appRoot.Id].ItemWithPath($"{folderName}/{fileName}").Content.PutAsync(ms)
                    logger.LogInformation("OneDrive upload completed for {FolderName}/{FileName}", folderName, fileName)
                    return (Ok "File uploaded")
            with
            | e -> 
                logger.LogError(e, "OneDrive upload failed for {FolderName}/{FileName}", folderName, fileName)
                return (Error e)
        }

File: onedrive.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
)

const (
	tokenURL    = "https://login.microsoftonline.com/common/oauth2/v2.0/token"
	graphAPIURL = "https://graph.microsoft.com/v1.0"
	chunkSize   = 320 * 1024
)

type tokenResponse struct {
	AccessToken  string `json:"access_token"`
	RefreshToken string `json:"refresh_token"`
}

type uploadSession struct {
	UploadURL string `json:"uploadUrl"`
}

func RefreshToken(clientID, refreshToken string) (accessToken, newRefreshToken string, err error) {
	form := url.Values{}
	form.Set("client_id", clientID)
	form.Set("grant_type", "refresh_token")
	form.Set("refresh_token", refreshToken)
	form.Set("redirect_uri", "http://localhost:8000")
	form.Set("scope", "Files.ReadWrite offline_access")

	resp, err := http.PostForm(tokenURL, form)
	if err != nil {
		return "", "", fmt.Errorf("token request: %w", err)
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)
	if resp.StatusCode != 200 {
		return "", "", fmt.Errorf("token request failed %d: %s", resp.StatusCode, body)
	}
	var tr tokenResponse
	if err := json.Unmarshal(body, &tr); err != nil {
		return "", "", fmt.Errorf("unmarshal token: %w", err)
	}
	return tr.AccessToken, tr.RefreshToken, nil
}

func UploadToOneDrive(accessToken, folderName, fileName string, data []byte) error {
	if len(data) < 4*1024*1024 {
		return simpleUpload(accessToken, folderName, fileName, data)
	}
	return chunkedUpload(accessToken, folderName, fileName, data)
}

func simpleUpload(accessToken, folderName, fileName string, data []byte) error {
	u := fmt.Sprintf("%s/me/drive/special/approot:/%s/%s:/content", graphAPIURL, folderName, fileName)
	req, err := http.NewRequest("PUT", u, bytes.NewReader(data))
	if err != nil {
		return fmt.Errorf("simple upload request: %w", err)
	}
	req.Header.Set("Authorization", "Bearer "+accessToken)
	req.Header.Set("Content-Type", "application/octet-stream")
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return fmt.Errorf("simple upload: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 300 {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("simple upload %d: %s", resp.StatusCode, body)
	}
	return nil
}

func chunkedUpload(accessToken, folderName, fileName string, data []byte) error {
	sessionURL := fmt.Sprintf("%s/me/drive/special/approot:/%s/%s:/createUploadSession", graphAPIURL, folderName, fileName)
	req, err := http.NewRequest("POST", sessionURL, strings.NewReader(`{"item":{"@microsoft.graph.conflictBehavior":"replace"}}`))
	if err != nil {
		return fmt.Errorf("create session request: %w", err)
	}
	req.Header.Set("Authorization", "Bearer "+accessToken)
	req.Header.Set("Content-Type", "application/json")
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return fmt.Errorf("create upload session: %w", err)
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)
	var session uploadSession
	if err := json.Unmarshal(body, &session); err != nil {
		return fmt.Errorf("unmarshal session: %w", err)
	}

	total := len(data)
	for offset := 0; offset < total; offset += chunkSize {
		end := offset + chunkSize
		if end > total {
			end = total
		}
		chunk := data[offset:end]
		chunkReq, err := http.NewRequest("PUT", session.UploadURL, bytes.NewReader(chunk))
		if err != nil {
			return fmt.Errorf("chunk request at %d: %w", offset, err)
		}
		chunkReq.Header.Set("Content-Range", fmt.Sprintf("bytes %d-%d/%d", offset, end-1, total))
		chunkReq.Header.Set("Content-Type", "application/octet-stream")
		chunkResp, err := http.DefaultClient.Do(chunkReq)
		if err != nil {
			return fmt.Errorf("upload chunk at %d: %w", offset, err)
		}
		io.Copy(io.Discard, chunkResp.Body)
		chunkResp.Body.Close()
		if chunkResp.StatusCode >= 400 {
			return fmt.Errorf("upload chunk %d: status %d", offset, chunkResp.StatusCode)
		}
	}
	return nil
}
